onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc

```diff
@@ -0,0 +1,242 @@
+            segment_col_index = segment_ids_data[index];
+            if (segment_col_index < 0 || segment_col_index >= segment_embedding_length) {
+              failed.store(true, std::memory_order_release);
+              return;
+            }
+          }
+
+          // Grab inputs for the embeddings for the current batch index:
+          const uint8_t* input_word_embedding = word_embedding_data + word_col_index * hidden_size;
+          const uint8_t* input_position_embedding =
+              position_embedding_data + position_col_index * hidden_size;
+          const uint8_t* input_segment_embedding = nullptr;
+          if (segment_embedding_data != nullptr) {
+            input_segment_embedding = segment_embedding_data + segment_col_index * hidden_size;
+          }
+
+          T* output = output_data + (index * hidden_size);
+
+          T sum = static_cast<T>(0);
+          for (int i = 0; i < hidden_size; ++i) {
+            T subtotal = Dequantize<uint8_t>(input_word_embedding[i],
+                                             word_embedding_scale,
+                                             word_embedding_zero_point) +
+                         Dequantize<uint8_t>(input_position_embedding[i],
+                                             position_embedding_scale,
+                                             position_embedding_zero_point);
+            if (segment_embedding_data != nullptr) {
+              subtotal += Dequantize<uint8_t>(input_segment_embedding[i],
+                                              segment_embedding_scale,
+                                              segment_embedding_zero_point);
+            }
+            output[i] = subtotal;
+            sum += subtotal;
+          }
+
+          T mean = sum / hidden_size;
+          sum = 0;
+
+          for (int i = 0; i < hidden_size; i++) {
+            T a = output[i] - mean;
+            output[i] = a;
+            sum += a * a;
+          }
+
+          T e = sqrt(sum / hidden_size + static_cast<T>(EmbedLayerNorm<T>::epsilon_));
+          for (int i = 0; i < hidden_size; i++) {
+            T cur_gamma = Dequantize<uint8_t>(gamma_data[i],
+                                              layer_norm_weights_scale,
+                                              layer_norm_weights_zero_point);
+            T cur_beta = Dequantize<uint8_t>(beta_data[i],
+                                             layer_norm_bias_scale,
+                                             layer_norm_bias_zero_point);
+            output[i] = output[i] / e * cur_gamma + cur_beta;
+          }
+        },
+        0);
+
+    if (failed.load(std::memory_order_acquire)) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "input index out of range");
+    }
+  }
+
+  // Calculate mask
+  if (nullptr != mask) {
+    const int32_t* mask_data = mask->template Data<int32_t>();
+    for (int b = 0; b < batch_size; b++) {
+      mask_index->template MutableData<int32_t>()[b] =
```

Comment:
Nit: probably slightly faster to move the `MutableData<>` call outside the loop. (I'm not sure if the compiler will do something appropriately intelligent with a macro that expands to `abort()` or throwing an exception.)